icicle/src/polynomials/cuda_backend/polynomial_cuda_backend.cu

```diff
@@ -0,0 +1,873 @@
+    }
+
+    void set_storage(void* storage, uint64_t nof_elements)
+    {
+      release();
+      m_storage = storage;
+      this->m_nof_elements = nof_elements;
+
+      modified();
+    }
+
+    // Note: this is protected and only backend can call
+    void* get_storage_mutable() override
+    {
+      // since giving access to internal memory, cannot know if modified or not
+      // backend should not take it mutable if not mutating
+      modified();
+      return m_storage;
+    }
+    const void* get_storage_immutable() override { return m_storage; }
+
+    void extend_mem_and_pad(uint64_t nof_elements)
+    {
+      void* new_storage = nullptr;
+      const uint64_t new_nof_elements = allocate_mem(nof_elements, &new_storage, true /*=memset zeros*/);
+      const uint64_t old_mem_size = this->m_nof_elements * ElementSize;
+
+      CHK_STICKY(
+        cudaMemcpyAsync(new_storage, m_storage, old_mem_size, cudaMemcpyDeviceToDevice, m_device_context.stream));
+
+      set_storage(new_storage, new_nof_elements);
+    }
+
+    void release() override
+    {
+      if (m_storage != nullptr) { CHK_STICKY(cudaFreeAsync(m_storage, m_device_context.stream)); }
+
+      m_storage = nullptr;
+      this->m_nof_elements = 0;
+
+      modified();
+    }
+
+    State get_state() const override { return m_state; }
+    void set_state(State state) { m_state = state; }
+    uint64_t get_nof_elements() const override { return m_nof_elements; }
+
+    bool is_host_ptr(const void* p) const
+    {
+      // checking whether a pointer is on host or device and asserts device matches the polynmoial device
+      // Note: device memory can me managed or not. Host memory can be registered or not. No distinction here.
+      cudaPointerAttributes attributes;
+      CHK_STICKY(cudaPointerGetAttributes(&attributes, p));
+      const bool is_on_host = attributes.type == cudaMemoryTypeHost ||
+                              attributes.type == cudaMemoryTypeUnregistered; // unregistered is host memory
+      const bool is_on_cur_device = !is_on_host && attributes.device == m_device_context.device_id;
+      const bool is_valid_ptr = is_on_host || is_on_cur_device;
+      if (!is_valid_ptr) { THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "Invalid evaluations ptr"); }
+
+      return is_on_host;
+    }
+
+    void from_coefficients(uint64_t nof_coefficients, const C* coefficients) override
+    {
+      const bool is_memset_zeros = coefficients == nullptr;
+      allocate(nof_coefficients, State::Coefficients, is_memset_zeros);
+      if (coefficients) {
+        const bool is_ptr_on_host = is_host_ptr(coefficients);
+
+        CHK_STICKY(cudaMemcpyAsync(
+          m_storage, coefficients, nof_coefficients * sizeof(C),
+          is_ptr_on_host ? cudaMemcpyHostToDevice : cudaMemcpyDeviceToDevice, m_device_context.stream));
+        CHK_STICKY(
+          cudaStreamSynchronize(m_device_context.stream)); // protect against coefficients being released too soon
```

Comment:
wdym by that? Why can't we let the user to decide if he wants async here?

Reply:
The problem I want to avoid is that the host data is still valid when copying.
I could give the user an 'async' flag so that he is aware that he must keep the data alive. The thing I don't like about that is that he would later need to call some blocking sync on the polynomial and it must be device agnostic.
So something like 

Polynomial p = Polynomial::from_coefficients(host_coeffs_p, nof_coeffs, true /*=async*/);
...
...
p.sync(); // or sync(p) or something like that
// can now release host_coeffs_p

We could do that yes. What do you think? I can also add it later since a last param that defaults to false will not break any user code.

Reply:
From our experience with zkwasm integration - async is a crucial part of achieving maximal throughput in applications. Even if we talk about ZPU, I think async will still be relevant.

But yeah, maybe we can add it later.